### PR TITLE
Weave images are always downloaded

### LINKED ISSUE #699: Weave images are always downloaded from Docker causing Docker Pull Rate error
Testing Submariner with generic `make deploy` from a server on corporate network and keep getting the Docker Pull Rate error on Weave Image:
```
docker: Error response from daemon: toomanyrequests: You have reached your pull rate limit.
  You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit. 
```
Weave is downloaded to host, but clusters in KIND deployment don't have access to it and are attempting to re-download it.

### PR DESCRIPTION
**Pull request by Billy99**
Weave is downloaded to host, but clusters in KIND deployment don't have
access to it and are attempting to re-download it. Update cluster.sh
such that it pre-downloads the Weave images and copies them to the local
registry so they are available within the KIND Cluster.

Closes: submariner-io/shipyard#699

Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

### CODE REVIEW

File: scripts/shared/clusters.sh [hunk 1]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
 
 function deploy_weave_cni(){
     echo "Applying weave network..."
-    curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!' | kubectl apply -f -
+
+    WEAVE_YAML=`curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!'`
+
+    # Search the YAML for images that need to be downloaded
+    IMAGE_LIST=( $(echo "${WEAVE_YAML}" | grep "image:" | awk -F"'" '{print $2}') )
```
Comment (skitt) [inline scripts/shared/clusters.sh]: The Shipyard base image has `yq`, please use that when extracting information from YAML. Note too that `awk` subsumes `grep`: `awk -F"'" '/image:/ {print $2}'`.

File: scripts/shared/clusters.sh [hunk 2]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
 
 function deploy_weave_cni(){
     echo "Applying weave network..."
-    curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!' | kubectl apply -f -
+
+    WEAVE_YAML=`curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!'`
```
Comment (skitt) [inline scripts/shared/clusters.sh]: Please avoid backticks for command substitution:

```suggestion
    WEAVE_YAML=$(curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!')
```

File: scripts/shared/clusters.sh [hunk 3]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+    # Search the YAML for images that need to be downloaded
+    IMAGE_LIST=( $(echo "${WEAVE_YAML}" | grep "image:" | awk -F"'" '{print $2}') )
+    echo "IMAGE_LIST=${IMAGE_LIST[@]}"
+    for i in "${!IMAGE_LIST[@]}"
+    do
+        IMAGE_FAILURE=false
+
+        # Check if image is already present, and if not, download it.
+        echo "Processing Image: ${IMAGE_LIST[$i]}"
+        if [ -z "`docker images -q ${IMAGE_LIST[$i]}`" ] ; then
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
        if [ -z "$(docker images -q $image)" ] ; then
```

File: scripts/shared/clusters.sh [hunk 4]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
 function deploy_weave_cni(){
     echo "Applying weave network..."
-    curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!' | kubectl apply -f -
+
+    WEAVE_YAML=`curl -sL "https://cloud.weave.works/k8s/net?k8s-version=v$k8s_version&env.IPALLOC_RANGE=${cluster_CIDRs[${cluster}]}" | sed 's!ghcr.io/weaveworks/launcher!weaveworks!'`
+
+    # Search the YAML for images that need to be downloaded
+    IMAGE_LIST=( $(echo "${WEAVE_YAML}" | grep "image:" | awk -F"'" '{print $2}') )
+    echo "IMAGE_LIST=${IMAGE_LIST[@]}"
+    for i in "${!IMAGE_LIST[@]}"
```
Comment (skitt) [inline scripts/shared/clusters.sh]: We don’t use the index, so we can iterate over the array contents directly:

```suggestion
    for image in "${IMAGE_LIST[@]}"
```

File: scripts/shared/clusters.sh [hunk 5]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+    do
+        IMAGE_FAILURE=false
+
+        # Check if image is already present, and if not, download it.
+        echo "Processing Image: ${IMAGE_LIST[$i]}"
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
        echo "Processing Image: $image"
```

File: scripts/shared/clusters.sh [hunk 6]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+        if [ -z "`docker images -q ${IMAGE_LIST[$i]}`" ] ; then
+            echo "Image ${IMAGE_LIST[$i]} not found, downloading..."
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
            echo "Image $image not found, downloading..."
```

File: scripts/shared/clusters.sh [hunk 7]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+            docker pull "${IMAGE_LIST[$i]}"
+            if [ $? != 0 ] ; then
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
            if ! docker pull "$image"; then
```

File: scripts/shared/clusters.sh [hunk 8]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+                echo "**** 'docker pull ${IMAGE_LIST[$i]}' failed. Manually run. ****"
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
                echo "**** 'docker pull $image' failed. Manually run. ****"
```

File: scripts/shared/clusters.sh [hunk 9]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+                IMAGE_FAILURE=true
+            fi
+        else
+            echo "Image ${IMAGE_LIST[$i]} already downloaded."
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
            echo "Image $image already downloaded."
```

File: scripts/shared/clusters.sh [hunk 10]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+        fi
+
+        if [ "${IMAGE_FAILURE}" == false ] ; then
+            LCL_REG_IMAGE_NAME=`echo "${IMAGE_LIST[$i]}" | sed 's!weaveworks/!localhost:5000/!'`
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
            LCL_REG_IMAGE_NAME="${image/weaveworks/localhost:5000}"
```

File: scripts/shared/clusters.sh [hunk 11]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+            # Copy image to local registry if not there
+            if [ -z "`docker images -q ${LCL_REG_IMAGE_NAME}`" ] ; then
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
            if [ -z "$(docker images -q ${LCL_REG_IMAGE_NAME})" ] ; then
```

File: scripts/shared/clusters.sh [hunk 12]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+                echo "Image ${LCL_REG_IMAGE_NAME} not found, tagging and pushing ..."
+                docker tag "${IMAGE_LIST[$i]}" "${LCL_REG_IMAGE_NAME}"
+                if [ $? != 0 ] ; then
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
                if ! docker tag "$image" "${LCL_REG_IMAGE_NAME}"; then
```

File: scripts/shared/clusters.sh [hunk 13]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+                    echo "'docker tag ${IMAGE_LIST[$i]} ${LCL_REG_IMAGE_NAME}' failed."
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
                    echo "'docker tag $image ${LCL_REG_IMAGE_NAME}' failed."
```

File: scripts/shared/clusters.sh [hunk 14]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+                    IMAGE_FAILURE=true
+                else
+                    docker push "${LCL_REG_IMAGE_NAME}"
+                    if [ $? != 0 ] ; then
```
Comment (skitt) [inline scripts/shared/clusters.sh]:
```suggestion
                    if ! docker push "${LCL_REG_IMAGE_NAME}"; then
```

File: scripts/shared/clusters.sh [hunk 15]
```diff
@@ -114,7 +114,57 @@ function deploy_cni() {
...
+                    fi
+                fi
+            else
+                echo "Image ${LCL_REG_IMAGE_NAME} already present."
+            fi
+        fi
+
+        if [ "${IMAGE_FAILURE}" == false ] ; then
+            # Update the YAML by replacing upstream image name with the local registry image name
+            WEAVE_YAML=`echo "${WEAVE_YAML}" | sed "s!image: '${IMAGE_LIST[$i]}'!image: '${LCL_REG_IMAGE_NAME}'!g"`
```
Comment (skitt) [inline scripts/shared/clusters.sh]: Please use `yq`.

**Review by skitt — CHANGES_REQUESTED** [review]
Thanks for taking care of this! I have a few suggestions...

File: scripts/shared/clusters.sh (hunk 2 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 4 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 5 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 3 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 6 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 7 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 8 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 9 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 10 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 11 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 12 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 13 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 14 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 1 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Done.

File: scripts/shared/clusters.sh (hunk 15 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: I spent half a day on yq, I can't figure out how to substitute. I only get LCL_REG_IMAGE_NAME to print, not the whole file with that substituted.

**Review by Billy99 — COMMENTED** [review]
Fixed all except using yq to replace the existing image name with the local registry name.

File: scripts/shared/clusters.sh (hunk 15 above)
Comment (Billy99) [reply scripts/shared/clusters.sh]: Another couple of hours and I figured it out.

**Comment by Billy99** [comment]
@skitt PTAL

**Review by dfarrell07 — APPROVED** [review]
LGTM, once @skitt re-reviews let's be sure to squash https://github.com/submariner-io/shipyard/pull/700/commits/e4f8e666ba442a208b32cb778f6893b917889807.

**Comment by Billy99** [comment]
@dfarrell07 Will do! Thanks for the time.

**Review by skitt — APPROVED** [review]

**Pull request merged**